votingsystem/src/general_meeting.rs

```diff
@@ -0,0 +1,271 @@
+        }
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+        tx.verify().map_err(|_| ExecuteError::InvalidSign);
+
+        let end_time = tx.tx.action.end_time;
+        let tallying = tx.tx.action.tallying_time;
+        let meeting = GeneralMeeting::new(end_time,tallying);
+        let key = meeting.id.as_ref();
+
+        self.storage_mut().set(key , serde_cbor::to_vec(&meeting).unwrap());
+        Ok(meeting.id)
+    }
+}
+
+impl Stateful for Context {
+    fn set_storage(&mut self, storage: ServiceRef<dyn SubStorageAccess>) {
+        self.storage.replace(storage.unwrap_import().into_remote());
+    }
+}
+
+impl InitGenesis for Context {
+    fn begin_genesis(&mut self) {}
+
+    fn init_genesis(&mut self, config: &[u8]) {
+        let admin: Public = serde_cbor::from_slice(&config).unwrap();
+        self.admin = Some(admin.clone());
+        self.storage_mut().set(admin.as_ref(), admin.rlp_bytes());
+    }
+
+    fn end_genesis(&mut self) {}
+}
+
+
+impl TxOwner for Context {
+    fn block_opened(&mut self, _: &Header) -> Result<(), HeaderError> {
+        Ok(())
+    }
+
+    fn execute_transaction(&mut self, transaction: &Transaction) -> Result<TransactionOutcome, ()> {
+        if let Err(error) = self.excute_tx(transaction) {
+            match error {
+                ExecuteError::InvalidMetadata => Err(()),
+                ExecuteError::InvalidSign => Err(()),
+                ExecuteError::InvalidFormat => Err(()),
+                ExecuteError::InvalidAdmin => Err(()),
+            }
+        } else {
+            Ok(Default::default())
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), coordinator::types::ErrorCode> {
+        let todo_fixthis: coordinator::types::ErrorCode = 3;
+        assert_eq!(transaction.tx_type(), "CreateGeneralMeeting");
+        let tx: OwnTransaction = serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+        tx.verify().map_err(|_| todo_fixthis)?;
+
+        let admin_public_key = tx.signer_public;
+        if !self.storage().has(admin_public_key.as_ref()) {
+           return Err(ExecuteError::InvalidAdmin).map_err(|_|todo_fixthis);
+        }
+
+        Ok(())
+    }
+
+    fn block_closed(&mut self) -> Result<Vec<Event>, CloseBlockError> {
+        Ok(Vec::new())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TimeStamp{
+    time: u64,
+}
+
+
+impl TimeStamp {
+    fn set_time(&mut self, time: u64){
+        self.time = time;
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct GeneralMeetingId {
+    id: u8,
+}
+
+impl GeneralMeetingId {
+    pub fn new ()->Self {
+        let mut rng = rand::thread_rng();
+        let random_id = rng.gen();
+        Self{
+            id: random_id
+        }
+    }
+}
+
+impl AsRef<[u8]> for GeneralMeetingId{
+    fn as_ref(&self)->&[u8]{
+        self.as_ref()
+    }
+}
```

Comment:
Did you add `as_ref` for the serialization? I think AsRef is not related to the serialization.
You can use `#[serde(transparent)]` to serialize a struct that has only a field exactly the same as if its one field were serialized itself.

https://serde.rs/container-attrs.html#transparent

I recommend you to read the serde document. It isn't long.

Reply:
OK, I understand your intention.

I thought you used as_ref for the value serialization.
How about naming the [u8] as `Key` to express your intention?

```rust
type StorageKeyRef<'a> = &'a [u8];

impl AsRef<StorageKey> for GeneralMeetingId{
    fn as_ref(&self)->StorageKeyRef {
      std::slice::from_ref(&self.id)
    }
}
```

Also calling `self.as_ref()` in the as_ref definition is a infinite loop. If you want to make &T to &[T] you should use `std::slice::from_ref`: https://doc.rust-lang.org/std/slice/fn.from_ref.html